### DISCUSSION TITLE
Flexible Model Selection Ideas

### DISCUSSION THREAD

**Discussion by a-priestley** [olimorris/codecompanion.nvim#91] [Ideas]
I'm not certain if there is already a way to switch adapters on the fly, but here is my clunky solution (depends on lazy):

```lua
local default_adapters = {
	chat = "openai",
	inline = "openai",
	agent = "openai",
}
local adapters = { "anthropic", "ollama", "openai" }
local function select_adapter(strategy)
	vim.ui.select(adapters, {
		prompt = "Select Adapter for " .. strategy .. " strategy:",
	}, function(choice)
		if choice then
			default_adapters[strategy] = choice
			require("lazy").reload({ plugins = { "codecompanion.nvim" } })
			vim.notify("Adapter for " .. strategy .. " strategy set to: " .. choice)
		end
	end)
end
vim.api.nvim_create_user_command("SelectChatAdapter", function()
	select_adapter("chat")
end, {})
vim.api.nvim_create_user_command("SelectInlineAdapter", function()
	select_adapter("inline")
end, {})
vim.api.nvim_create_user_command("SelectAgentAdapter", function()
	select_adapter("agent")
end, {})
return {
	"olimorris/codecompanion.nvim",
	dependencies = {
		"nvim-lua/plenary.nvim",
		"nvim-treesitter/nvim-treesitter",
		"nvim-telescope/telescope.nvim", -- Optional
		{
			"stevearc/dressing.nvim", -- Optional: Improves the default Neovim UI
			opts = {},
		},
	},
	config = function()
		local codecompanion = require("codecompanion")
		codecompanion.setup({
			adapters = {
				anthropic = "anthropic",
				ollama = "ollama",
				openai = "openai",
			},
			strategies = {
				-- CHAT STRATEGY ----------------------------------------------------------
				chat = {
					adapter = default_adapters.chat,
					roles = {
						llm = "CodeCompanion", -- The markdown header content for the LLM's responses
						user = "Me", -- The markdown header for your questions
					},

-- rest of config
...
```

The extra clunky bit is where the plugin is completely reloaded to get the selection to take effect, but maybe there is already functionality in codecompanion to refresh that specific piece of the config?

Let me know!

**Comment by olimorris** [comment]
Love this and thanks for sharing!!

I really should make it easier to switch out an adapter. I'll put that on the backlog!

**Reply by olimorris** [reply]
@a-priestley I've just pushed a feature where you can now switch adapters in the chat buffer by pressing `ga`. Let me know what you think.

Do you think selecting the model would be useful as well?

**Reply by a-priestley** [reply]
This is great! is this only usable in chat though? For me it is more useful to allow setting for each defined strategy individually. Personally I think It would be best as a CodeCompanion actions item.
As for model switching, I think that is already available but again only in chat. I wonder if maybe we should be thinking about models per strategy as opposed to adapters per strategy. Maybe like a drill down? Eg, `CodeCompanion actions` > `Switch Model for [Strategy]` > `Adapter` > `Model`? 
While on the topic of model switching, I'm wondering if you are hard coding the model tables? When I am switching models in chat, I can see only a few in the list when I know there are many more. In case you are not aware, the list of available models can be fetched from both ollama and openai, although the last time I checked anthropic lacked this API endpoint.

**Reply by SDGLBL** [reply]
I agree, implementing this as a CodeCompanion actions item makes a lot of sense. It's a solid suggestion that could really improve our workflow. Being able to set up more cost-effective models like GPT-4-mini for specific inline actions would give us useful flexibility. For instance, we could use a lighter model for tasks like writing code comments. This would be perfect for quickly generating docstrings or inline comments, where we don't necessarily need the full power of a larger model.

**Reply by a-priestley** [reply]
I think that's reasonable! It seems like you're suggesting an even more fine-grained approach? I hadn't considered model selection for individual prompts -- just the three overarching strategies.
Model per prompt might be better as a configuration item imo:

```lua
["Generate a Commit Message"] = {
	strategy = "chat",
	description = "Generate a commit message",
	opts = {
		index = 5,
		-- NEW OPTION
		model = "gpt-3.5-turbo",
		--
		default_prompt = true,
		mapping = "<LocalLeader>cm",
		slash_cmd = "commit",
		auto_submit = true,
	},
	prompts = {
		{
			role = "${user}",
			contains_code = true,
			content = function()
				return "You are an expert at following the Conventional Commit specification. Given the git diff listed below, please generate a commit message for me:"
					.. "\n\n```\n"
					.. vim.fn.system("git diff --staged")
					.. "\n```"
			end,
		},
	},
},
```

On the other hand, could do something in addition to this for prompting the user for model selection when calling for the prompt:
```lua
["Generate a Commit Message (Model Select)"] = {
	strategy = "chat",
	description = "Generate a commit message",
	opts = {
		index = 5,
		-- NEW OPTION
		default_model = "gpt-3.5-turbo",
		select_model = true,
		--
		default_prompt = true,
		mapping = "<LocalLeader>cm",
		slash_cmd = "commit",
		auto_submit = true,
	},
	prompts = {
		{
			role = "${user}",
			contains_code = true,
			content = function()
				return "You are an expert at following the Conventional Commit specification. Given the git diff listed below, please generate a commit message for me:"
					.. "\n\n```\n"
					.. vim.fn.system("git diff --staged")
					.. "\n```"
			end,
		},
	},
},
```

Having `select_model = true` might bring up a table of available models, with the value assigned to `default_model` appearing at the top of the list.

**Reply by olimorris** [reply]
Great suggestions. I've finished the ability to select an adapter and then a model in the chat buffer so can leverage some of that for the inline adapter.

Something like:

```lua
{
  opts = {
    adapter = {
      name = "openai",
      model = "gpt-4o-mini"
    }
  }
}
```

Seems sensible and would help easy to implement.

You could also pass values into an inline prompt with something like `:CodeCompanion /buffer write some unit tests for this file --adapter=openai --gpt-4o-mini`. But I get the feeling not many people use that command.

**Reply by olimorris** [reply]
Hi Both. Commit f95efb3b781edfaeb3069d791c59451d0b09aa57 includes this functionality. Keen to get your feedback from initial testing particularly with inline prompts

**Reply by SDGLBL** [reply]
Works well for me

**Reply by a-priestley** [reply]
Is this a breaking change?

```
Error executing vim.schedule lua callback: vim/keymap.lua:0: rhs: expected string|function, got nil
stack traceback:
	[C]: in function 'error'
	vim/shared.lua: in function 'validate'
	vim/keymap.lua: in function 'set'
	...y/codecompanion.nvim/lua/codecompanion/utils/keymaps.lua:61: in function 'set'
	...codecompanion.nvim/lua/codecompanion/strategies/chat.lua:377: in function 'open'
	...codecompanion.nvim/lua/codecompanion/strategies/chat.lua:294: in function ''
	vim/_editor.lua: in function <vim/_editor.lua:0>
```
This is happening when trying to open the basic chat action.

I may just be confused as to where to place the new opts property?

**Reply by a-priestley** [reply]
I think I found the culprit.
```lua
-- DISPLAY OPTIONS ----------------------------------------------------------
display = {
  action_palette = {
	  width = 95,
	  height = 10,
  },
  chat = {
	  window = {
		  layout = "vertical", -- float|vertical|horizontal|buffer
		  border = "single",
		  height = 0.8,
		  width = 0.45,
		  relative = "editor",
		  opts = {
			  breakindent = true,
			  cursorcolumn = false,
			  cursorline = false,
			  foldcolumn = "0",
			  linebreak = true,
			  list = false,
			  signcolumn = "no",
			  spell = false,
			  wrap = true,
			  -- TAKE THIS OUT
			  -- adapter = {
			  -- 	name = "openai",
			  -- 	model = "gpt-4o",
			  -- },
		  },
	  },
	  intro_message = "Welcome to CodeCompanion ✨! Press ? for help",
  
	  messages_separator = "─", -- The separator between the different messages in the chat buffer
	  show_separator = true, -- Show a separator between LLM responses?
	  show_settings = true, -- Show LLM settings at the top of the chat buffer?
	  show_token_count = true, -- Show the token count for each response?
  },
-- rest of config
```

**Reply by olimorris** [reply]
Check out the commit that added the functionality: https://github.com/olimorris/codecompanion.nvim/commit/f95efb3b781edfaeb3069d791c59451d0b09aa57

I covered it in RECIPES.md.

**Reply by a-priestley** [reply]
Yep, I've got it now. Thanks!! :+1:

**Reply by a-priestley** [reply]
I've come up with yet more hacky code to get the plugin to do what I want.
Basically, the initial plugin opts are loaded into a table external to the plugin config function itself. The `select_model()` function takes a user selection for the desired adapter, followed by a selection for that adapters available models. Given the select, we update the appropriate values in the opts table, and completely reload the plugin using the `setup()` func (I couldn't figure out how to do this on the plugins terms outside of the chat buffer keymap).
For now I took out the "adapter per strategy" just because I think it's easier to keep just one model selection in my head at a time while I try out different ones for different purposes. 
I wanted to have this as a CodeCompanionActions item, but I'm not very good at lua yet. There's a circular dependency issue there, where the opts table needs to be defined before the select_model function, since the latter is mutating the former.

```lua
local adapters = {
	anthropic = { name = "anthropic", model = "claude-3-5-sonnet-20240620" },
	ollama = { name = "ollama", model = "llama:3.1:8b" },
	openai = { name = "openai", model = "gpt-4o" },
	gemini = { name = "gemini", model = "gemini-1.5-flash" },
}
local default_adapter = adapters.ollama

local opts = {
	adapters = {
		ollama = function()
			return require("codecompanion.adapters").extend("ollama", {
				name = "ollama",
				schema = {
					model = {
						default = adapters.ollama.model,
					},
				},
			})
		end,
                --etc
        },
	strategies = {
		-- CHAT STRATEGY ----------------------------------------------------------
		chat = {
			adapter = default_adapter.name,
			--etc
		},
	        -- etc
	},
	-- etc
}

---Select the default model
local function select_model()
	vim.ui.select(vim.tbl_keys(adapters), {
		prompt = "Select Adapter:",
	}, function(adapter_choice)
		if not adapter_choice then
			return
		end
		local model_choices = {}
		model_choices = require("codecompanion.adapters." .. adapter_choice).schema.model.choices
		if type(model_choices) == "function" then
			model_choices = model_choices()
		end
		vim.ui.select(model_choices, {
			prompt = "Select Model:",
		}, function(model_choice)
			if model_choice then
				adapters[adapter_choice].model = model_choice
				default_adapter = adapters[adapter_choice]
				opts.strategies.chat.adapter = default_adapter.name
				opts.strategies.inline.adapter = default_adapter.name
				opts.strategies.agent.adapter = default_adapter.name
				require("codecompanion").setup(opts)
			end
		end)
	end)
end

vim.api.nvim_create_user_command("SelectModel", function()
	select_model()
end, {})

return {
	"olimorris/codecompanion.nvim",
	dependencies = {
		"nvim-lua/plenary.nvim",
		"nvim-treesitter/nvim-treesitter",
		"nvim-telescope/telescope.nvim", -- Optional
		"stevearc/dressing.nvim", -- Optional: Improves the default Neovim UI
	},
	config = function()
		local codecompanion = require("codecompanion")
		codecompanion.setup(opts)
                -- etc
	end,
}
```
